File: Resources/WebResources/common/ApplicationSettings.js
/*
 * Copyright (c) 2022 Maximilian Bauer
 * Distributed under the GNU GPL v3. For full terms see the file "COPYING".
 */
// @ts-check

import { AutomaticWateringSettings } from "./AutomaticWateringSettings.js";
import { HardwareSettings } from "./HardwareSettings.js";
import { MoistureWarningSettings } from "./MoistureWarningSettings.js";
import { Settings } from "./Settings.js";
import { TemperatureWarningSettings } from "./TemperatureWarningSettings.js";
import { WaterFillLevelWarningSettings } from "./WaterFillLevelWarningSettings.js";

export class ApplicationSettings extends Settings {
   /** @type {AutomaticWateringSettings} */
   automaticWateringSettings;
   /** @type {MoistureWarningSettings} */
   moistureWarningSettings;
   /** @type {TemperatureWarningSettings} */
   temperatureWarningSettings;
   /** @type {WaterFillLevelWarningSettings} */
   waterFillLevelWarningSettings;
   /** @type {HardwareSettings} */
   hardwareSettings;

   /**
    * @param {AutomaticWateringSettings} automaticWateringSettings
    * @param {MoistureWarningSettings} moistureWarningSettings
    * @param {TemperatureWarningSettings} temperatureWarningSettings
    * @param {WaterFillLevelWarningSettings} waterFillLevelWarningSettings
    * @param {HardwareSettings} hardwareSettings
    */
   constructor(automaticWateringSettings, moistureWarningSettings,
      temperatureWarningSettings, waterFillLevelWarningSettings, 
      hardwareSettings) {
      super();
      this.automaticWateringSettings = automaticWateringSettings;
      this.moistureWarningSettings = moistureWarningSettings;
      this.temperatureWarningSettings = temperatureWarningSettings;
      this.waterFillLevelWarningSettings = waterFillLevelWarningSettings;
      this.hardwareSettings = hardwareSettings;
   }

   clone() {
      return new ApplicationSettings(this.automaticWateringSettings, this.moistureWarningSettings,
         this.temperatureWarningSettings, this.waterFillLevelWarningSettings, 
         this.hardwareSettings);
   }
}

File: GrowOne.Core/Settings/ApplicationSettings.cs
/*
 * Copyright (c) 2022 Maximilian Bauer
 * Distributed under the GNU GPL v3. For full terms see the file "COPYING".
 */
#nullable enable

namespace GrowOne.Core.Settings
{
    public class ApplicationSettings
    {
        public AutomaticWateringSettings? AutomaticWateringSettings { get; set; }

        public MoistureWarningSettings? MoistureWarningSettings { get; set; }

        public TemperatureWarningSettings? TemperatureWarningSettings { get; set; }

        public WaterFillLevelWarningSettings? WaterFillLevelWarningSettings { get; set; }

        public HardwareSettings? HardwareSettings { get; set; }

        public static ApplicationSettings CreateDefault()
        {
            return new ApplicationSettings()
            {
                AutomaticWateringSettings = new AutomaticWateringSettings()
                {
                    Enabled = false,
                    CooldownSeconds = 900,
                    DurationSeconds = 10,
                    MinimumMoisture = 0.5f
                },
                MoistureWarningSettings = new MoistureWarningSettings()
                {
                    Enabled = false,
                    MaximumMoisture = 1,
                    MinimumMoisture = 0.4f
                },
                TemperatureWarningSettings = new TemperatureWarningSettings()
                {
                    Enabled = false,
                    MaximumTemperature = 28,
                    MinimumTemperature = 17
                },
                WaterFillLevelWarningSettings = new WaterFillLevelWarningSettings()
                {
                    Enabled = false,
                    MinimumLevel = 0.25f,
                    MaximumLevel = 1
                },
                HardwareSettings = new HardwareSettings()
            };
        }

        public ApplicationSettings Clone()
        {
            return new ApplicationSettings()
            {
                AutomaticWateringSettings = AutomaticWateringSettings?.Clone(),
                MoistureWarningSettings = MoistureWarningSettings?.Clone(),
                TemperatureWarningSettings = TemperatureWarningSettings?.Clone(),
                WaterFillLevelWarningSettings = WaterFillLevelWarningSettings?.Clone(),
                HardwareSettings = HardwareSettings?.Clone()
            };
        }
    }
}
